build_tools/cmake/iree_hal_cts_test_suite.cmake

```diff
@@ -30,18 +39,92 @@ function(iree_hal_cts_test_suite)
   cmake_parse_arguments(
     _RULE
     ""
-    "DRIVER_NAME;DRIVER_REGISTRATION_HDR;DRIVER_REGISTRATION_FN"
+    "DRIVER_NAME;DRIVER_REGISTRATION_HDR;DRIVER_REGISTRATION_FN;COMPILER_TARGET_BACKEND;EXECUTABLE_FORMAT"
     "DEPS;EXCLUDED_TESTS;LABELS"
     ${ARGN}
   )
 
+  # Omit tests for which the specified driver is not enabled.
+  string(TOUPPER ${_RULE_DRIVER_NAME} _UPPERCASE_DRIVER)
+  string(REPLACE "-" "_" _NORMALIZED_DRIVER ${_UPPERCASE_DRIVER})
+  if(NOT DEFINED IREE_HAL_DRIVER_${_NORMALIZED_DRIVER})
+    message(SEND_ERROR "Unknown driver '${_RULE_DRIVER_NAME}'. Check IREE_HAL_DRIVER_* options.")
+  endif()
+  if(NOT IREE_HAL_DRIVER_${_NORMALIZED_DRIVER})
+    return()
+  endif()
+
   list(APPEND _RULE_LABELS "driver=${_RULE_DRIVER_NAME}")
 
+  # Enable executable tests if a compiler target backend capable of producing
+  # executables compatible with the driver is provided and enabled.
+  set(_ENABLE_EXECUTABLE_TESTS OFF)
+  if(DEFINED _RULE_COMPILER_TARGET_BACKEND)
+    string(TOUPPER ${_RULE_COMPILER_TARGET_BACKEND} _UPPERCASE_TARGET_BACKEND)
+    string(REPLACE "-" "_" _NORMALIZED_TARGET_BACKEND ${_UPPERCASE_TARGET_BACKEND})
+    if(NOT DEFINED IREE_TARGET_BACKEND_${_NORMALIZED_TARGET_BACKEND})
+      message(SEND_ERROR "Unknown backend '${_RULE_COMPILER_TARGET_BACKEND}'. Check IREE_TARGET_BACKEND_* options.")
+    endif()
+    if(DEFINED IREE_HOST_BINARY_ROOT)
+      # If we're not building the host tools from source under this configuration,
+      # such as when cross compiling, then we can't easily check for which
+      # compiler target backends are enabled. Just assume all are enabled and only
+      # rely on the runtime HAL driver check above for filtering.
+      set(_ENABLE_EXECUTABLE_TESTS ON)
+    else()
+      # We are building the host tools, so check enabled compiler target backends.
+      if(IREE_TARGET_BACKEND_${_NORMALIZED_TARGET_BACKEND})
+        set(_ENABLE_EXECUTABLE_TESTS ON)
+      endif()
+    endif()
+  endif()
+
+  # Generate testdata if executable tests are enabled.
+  if(_ENABLE_EXECUTABLE_TESTS)
+
+    set(_TRANSLATE_FLAGS
+      "-iree-mlir-to-hal-executable"
+      "-iree-hal-target-backends=${_RULE_COMPILER_TARGET_BACKEND}"
+    )
+    if(ANDROID)
+      set(_TARGET_TRIPLE "aarch64-none-linux-android${ANDROID_PLATFORM_LEVEL}")
+      list(APPEND _TRANSLATE_FLAGS "--iree-llvm-target-triple=${_TARGET_TRIPLE}")
+    endif()
+
+    # Skip over if already created (multiple drivers using the same compiler).
+    iree_package_name(_PACKAGE_NAME)
+    if(NOT TARGET ${_PACKAGE_NAME}_abs_executable_${_RULE_COMPILER_TARGET_BACKEND}_c)
```

Comment:
Hardcoding `abs` is not great :/

We are going to end up with a lot more than just abs - some of which we can do in a single executable with multiple entry points (variants with/without push constants, various descriptor set layouts, etc) and others that we'll need to do separately as support will vary (i64/f64 datatypes, etc). What we definitely don't want is to have to copy/paste dozens of lines in here.

Instead of having a generic `abs` that we randomly use we should have executables per functional test - that way changing one for a test doesn't break all other tests that are also relying on it. The test of whether or not push_descriptor_set works should not share the same mondo-executable used for testing push constants and indirect dispatch, etc. What data the executable produces is also going to change for different tests - having a push_constants test that writes the provided constants to the output buffer which can then be read back on the host, for example.

Would much rather have 20 files that have a bit of overlap than 2 files that are tangled through all the tests and impossible to change.

c_embed_data can have multiple files in it - compile each, embed as a single unit, and then at runtime have the tests lookup their file by name in the toc with a `get_test_executable_data("command_buffer_push_constants.bin")` etc

Reply:
Thanks for pushing back on this. I reworked into a single c_embed_data that can wrap multiple executables. PTAL.

Reply:
Yeah thank you Ben. I thought the use of abs was a bit weird, but assumed that it was going to be the test case for everything or something.